tests/test_driver.sh

```diff
@@ -0,0 +1,130 @@
+IFS="$(printf "\n\b")"
+
+# The first argument is the path to the script to launch; all remaining arguments are forwarded to
+# the script.
+fish_script="$1"
+shift 1
+script_args="${@}"
+
+die() {
+    if test "$#" -ge 0; then
+        printf "%s\n" "$@" 1>&2
+    fi
+    exit 1
+}
+
+# To keep things sane and to make error messages comprehensible, do not use relative paths anywhere
+# in this script. Instead, make all paths relative to one of these or $homedir."
+TESTS_ROOT="$(dirname "$0")"
+BUILD_ROOT="${TESTS_ROOT}/.."
+FISH_ROOT="${BUILD_ROOT}/.."
+
+# macOS (still) doesn't have `readlink -f` or `realpath`. It's OK, this is just for aesthetics.
+if which realpath 1>/dev/null 2>/dev/null; then
+    TESTS_ROOT="$(realpath "${TESTS_ROOT}")"
+    BUILD_ROOT="$(realpath "${BUILD_ROOT}")"
+    FISH_ROOT="$(realpath "${FISH_ROOT}")"
+fi
+
+if test -z "$__fish_is_running_tests"; then
+    # Set up a test environment and re-run the original script. We do not share environments
+    # whatsoever between tests, so each test driver run sets up a new profile altogether.
+
+    # macOS 10.10 requires an explicit template for `mktemp` and will create the folder in the
+    # current directory unless told otherwise. Linux isn't guaranteed to have $TMPDIR set.
+    homedir="$(mktemp -d 2>/dev/null || mktemp -d "${TMPDIR}tmp.XXXXXXXXXX")"
+
+    XDG_DATA_HOME="$homedir/xdg_data_home"
+    export XDG_DATA_HOME
+    mkdir -p $XDG_DATA_HOME/fish || die
+
+    XDG_CONFIG_HOME="$homedir/xdg_config_home"
+    export XDG_CONFIG_HOME
+    mkdir -p $XDG_CONFIG_HOME/fish || die
+
+
+    XDG_RUNTIME_DIR="$homedir/xdg_runtime_dir"
+    export XDG_CONFIG_HOME
+    mkdir -p $XDG_RUNTIME_DIR/fish || die
+
+    # Create a temp/scratch directory for tests to use, if they want (tests shouldn't write to a
+    # shared temp folder).
+    TMPDIR="$homedir/temp"
+    mkdir ${TMPDIR}
+    export TMPDIR
+
+    # These are used read-only so it's OK to symlink instead of copy
+    rm -f "$XDG_CONFIG_HOME/fish/functions"
+    ln -s "$PWD/test_functions" "$XDG_CONFIG_HOME/fish/functions" || die "Failed to symlink"
+
+    # Set the function path at startup, referencing the default fish functions and the test-specific
+    # functions.
+    fish_init_cmd="set fish_function_path ${XDG_CONFIG_HOME}/fish/functions ${FISH_ROOT}/share/functions"
+
+    __fish_is_running_tests="$homedir"
+    export __fish_is_running_tests
+
+    # Set locale information for consistent tests. Fish should work with a lot of locales but the
+    # tests assume an english UTF-8 locale unless they explicitly override this default. We do not
+    # want the users locale to affect the tests since they might, for example, change the wording of
+    # logged messages.
+    #
+    # TODO: set LANG to en_US.UTF-8 so we test the locale message conversions (i.e., gettext).
+    unset LANGUAGE
+    # Remove "LC_" env vars from the test environment
+    for kv in $(env); do
+        if echo "$kv" | grep -Eq "^LC_"; then
```

Comment:
This executes one grep per line - why not `$(env | grep)` above?

(yes, technically variables can contain newlines, but then you can't use `env` to begin with)

Reply:
Alternatively we could just hardcode the LC_ variables we know:

LANG
LC_CTYPE
LC_NUMERIC
LC_TIME
LC_COLLATE
LC_MONETARY
LC_MESSAGES
LC_PAPER
LC_NAME
LC_ADDRESS
LC_TELEPHONE
LC_MEASUREMENT
LC_IDENTIFICATION
LC_ALL

I don't think anyone adds anything more to that silly broken system?

Reply:
Thanks. I remember having a particular stupid moment with that bit of code, for some reason I was doing something incredibly convoluted based off of trying to port the previous logic line-by-line from fish and not accounting for the differences between `set` and `env`. The entire thing (including both `grep` invocations) can be executed just once.